DESIGN.md

```diff
@@ -0,0 +1,205 @@
+- Payload
+    Bytes that can be passed along with any wormhole message that contain application-specific data
+- Finality/Finality Threshold
+    The required number of blocks to wait until a VAA is produced
+
+# Discussion
+
+
+## What's the purpose of the Wormhole class?
+
+Wormhole class provides the main interface to do _everything_
+
+- Registers Platforms to access later -- constructor
+- Provides access to PlatformContexts -- getContext(ChainName)
+- Provides "shortcuts" to start a WormholeTransfer -- tokenTransfer/nftTransfer/cctp/...
+- Helpers for getting VAAs? or generally querying the API?
+- Abstract away chain-specific logic for easy mode access to methods
+
+## What do we want from a PlatformContext and how is that different from a provider / common utilities for a given platform?
+
+Provides Platform specific logic for a set of things
+
+- Register Modules (contract/app specific functionality)
+- Translates Platform specific stuff to generic stuff (e.g. ethers.Provider => RPC connection)
+- Deals with Platform specific interaction w/ chain (approve on eth, postvaa on sol, ...)
+- Implements standardized method format
+
+## What's the relationship between platforms/chains/providers?
+
+- A Platform provides the logic for all chains that run on that platform
+- A Chain provides consts (rpc/contract addresses/chain specific overrides)
+- A Provider is just an RPC connection and is held by the Chain. Providers are an implementation detail.
+
+## What's a signer vs. a wallet? Should signers have a provider (my answer: no)?
+
+- A Signer is an interface to sign transactions
+- It _may_ be backed by a wallet but not necessarily, as long as it fulfils the interface
+
+## Can we provide some way to make other non-standard applications available to use through the WormholeTransfer?
+
+Say I have an app that defines its own protocol, can I provide something that adheres to the WormholeTransfer interface so a dev can install it and call it like the TokenTransfer?
+
+
+# Outstanding Questions: 
+
+What is the preferred terminology to refer to either end of a cross-chain message: from/to, source/target or origin/destination?
+
+What is the preferred terminology for the core Wormhole layer? (i.e. Core Contracts or Wormhole Contracts)
+
+Should we namespace export base/definitions? 
+
+How should we think about xchain concepts without having xchain context?
+
+    e.g.  
+    // need to implement universal univeral decoder?
+    given eth address, and without installing evm platform, how do i turn it into a solana wrapped token without knowing how to fmt the address? 
+
+    // neet to tweak contracts on sol to support !ATA?
+    given an xfer from eth=>sol, and without installing sol platform, how do i determine the ATA?
```

Comment:
yeah, that one is a tricky bitch. I realized that this would be an issue a while ago but haven't put a lot of thought into it myself yet.

My knee-jerk response is to split platforms into a thin layer that contains light-weight stuff (like address types) and a heavy weight part (if you actually want to do stuff on that chain). But I'm afraid that this is not an abstraction that will actually work/hold up. (If you have to do address translations via RPC lookups like on Sui, you're presumably fucked).